### ISSUE TITLE
Build error with clang 11 on CUDA backend

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#3553]
This code can't be built with clang 11 with CUDA 10.1.243:
```c++
#include <Kokkos_Core.hpp>
#include <Kokkos_DualView.hpp>
#include <Kokkos_Random.hpp>
using ExecutionSpace = Kokkos::DefaultExecutionSpace;
using namespace std;

int main(int argc, char* argv[]) {
  Kokkos::initialize(argc,argv); {

  Kokkos::DualView<int*,ExecutionSpace> a("a",10);
  Kokkos::Random_XorShift64_Pool<ExecutionSpace> rndPool(1931);
  Kokkos::fill_random(a.d_view,rndPool,a.extent(0));

  Kokkos::DualView<int*,ExecutionSpace> aCum("aCum",a.extent(0));
  Kokkos::parallel_scan(a.extent(0),KOKKOS_LAMBDA(const int& i,int& update,const bool& final) {
    update += a.d_view(i);
    if(final) { aCum.d_view(i) = update; }
  });
  Kokkos::fence();

  } Kokkos::finalize();
}

```

which gives the error:

```c++
/home/aznb/mycodes/testKokkos/build_cuda_clang/_deps/kokkos-src/core/src/impl/Kokkos_FunctorAdapter.hpp:1411:14: error: reference to __host__ function 'operator new' in __host__ __device__ function
    return *(new (p) T());
             ^
/home/aznb/mycodes/testKokkos/build_cuda_clang/_deps/kokkos-src/core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2007:16: note: called by 'initial'
    ValueInit::init(m_functor, shared_value);
               ^
/home/aznb/mycodes/testKokkos/build_cuda_clang/_deps/kokkos-src/core/src/Cuda/Kokkos_Cuda_Parallel.hpp:2131:9: note: called by 'operator()'
        initial();
        ^
/home/aznb/mycodes/testKokkos/build_cuda_clang/_deps/kokkos-src/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp:116:3: note: called by 'cuda_parallel_launch_local_memory<Kokkos::Impl::ParallelScan<(lambda at /home/aznb/mycodes/testKokkos/testScan.cpp:26:37), Kokkos::RangePolicy<Kokkos::Cuda>, Kokkos::Cuda>>'
  driver();
  ^
/home/linuxbrew/.linuxbrew/opt/gcc@9/include/c++/9.3.0/new:173:33: note: 'operator new' declared here
_GLIBCXX_NODISCARD inline void* operator new(std::size_t, void* __p) _GLIBCXX_USE_NOEXCEPT

```

**Comment by crtrott** [comment]
looks like a compiler issue. Note this is a placement new, that SHOULD work on the device.

**Comment by crtrott** [comment]
A fix for this may be a bit longer, i.e. us trying to figure out how to do side step that.

**Comment by dalg24** [comment]
Check if that is still an issue and if that was fixed in more recent versions.

**Comment by ldh4** [comment]
I'm unable to reproduce this build error.
I was able to build and run that code with `Clang 11 + Cuda 10.1.243` and with `Clang 12 + Cuda 11.2.1` on both Kokkos `develop` and `master` (release 3.5) branch.

**Comment by ldh4** [comment]
Tried to this again with `Clang 11 + Cuda 10.1.243` with standard lib from `gcc 9.3.0`, but the code builds and runs for me.

**Comment by crtrott** [comment]
@Char-Aznable can you check whether this is still a problem?

**Comment by Char-Aznable** [comment]
I currently don't have a GPU node with CUDA 10 to test this. I believe this not likely persists if @ldh4 can reproduce the error with the Kokkos develop version around the OP's date (Oct 30, 2020) and see the error gone for the current version

**Comment by ldh4** [comment]
@Char-Aznable, I tried compiling a couple commits around Oct 30, 2020 from Kokkos develop branch, but I was still unable to reproduce this compilation error. Do you remember whether you used the `nvcc_wrapper` or not when compiling this?

**Comment by Char-Aznable** [comment]
@ldh4 I didn't use `nvcc_wrapper` but just clang++ for this. Here's what my cmake command looks like for building this:
```cmake
cmake .. \
-DCMAKE_CXX_COMPILER=$CLANGBIN/clang++ \
-DCMAKE_CXX_STANDARD=14 \
-DCMAKE_BUILD_TYPE=Release \
-DCMAKE_PREFIX_PATH=$BREW \
-DKokkos_ENABLE_CUDA=ON \
-DKokkos_ARCH_SNB=ON \
-DKokkos_ARCH_${cudaArch}=ON \
-DKokkos_ENABLE_CUDA_LAMBDA=ON \
-DKokkos_ENABLE_DEPRECATED_CODE=OFF \
-DKokkos_ENABLE_EXPLICIT_INSTANTIATION=OFF

```
where `$BREW` is the `/home/linuxbrew/.linuxbrew` in the error log message. `$CLANGBIN` is `$BREW/bin`.

**Comment by ldh4** [comment]
@Char-Aznable I tried a couple different configurations using your cmake command, but still no luck in reproducing this compilation error. I mostly used your cmake command as is, but I specified `--gcc-toolchain` to include gcc 9.3 files over what's found in my sysroot by default. Separately, I also compiled this with a locally built `clang11`, just in case the module installed on the machine I used had some unique configuration, but that didn't make any difference either.
